ls1intum/Artemis
# [Bugfix] Fix Back Button

### Checklist

### Motivation and Context
Resolves part of https://github.com/ls1intum/Orion/issues/50

### Description
The back-button just used location.back, which doesn't work if there is no previous page e.g. because Orion directly loads the correct link without entering the overview.

Artemis does have a helper to solve this issue, however the helper is not working correctly because it uses history.length to check if the page is the first page which does not work reliably since Firefox counts from 1, while Chrome counts from 2 (Chrome also counts the empty page). As discussed in great length [here](https://stackoverflow.com/questions/3588315/how-to-check-if-the-user-can-go-back-in-browser-history-or-not), there is no reliable solution. Thefore I added a new service that tracks the internal navigation, being independent from javascript primitives.

### Steps for Testing
1. Log in to Artemis
2. Create/Locate an exercise as student
3. Click on the back button
![grafik](https://user-images.githubusercontent.com/61357727/123513429-81b83d00-d68d-11eb-9ff2-248822dc73f8.png)
4. You should end up on the previous page
5. Navigate to the exercise from a different source, e.g. by clicking on the diagram points on the statistics page
![grafik](https://user-images.githubusercontent.com/61357727/123513469-b88e5300-d68d-11eb-88ee-20879286d7ef.png)
6. The back button should correctly lead you back to the different previous page
7. Copy and paste the url of the programming exercise into a new window/tab
8. The back button should now lead to the exercise overview
9. Also check the other buttons by copy/pasting the url and making sure it is still returned correctly. See changes for details.
    1. Save/Cancel after creating/editing a system notification
    2. Save/Cancel after creating/editing a user
    3. Save/Cancel after creating/editing a course
    4. Save/Cancel after creating/editing an exam
    5. Save/Cancel after creating/editing an exercise hint
    6. Save/Cancel after creating/editing a lecture
10. Optional: Try out different browsers

### Test Coverage
navigation.utils.ts 73.8% -> 82.1%
